src/Nethermind/Nethermind.Network/TrustedNodes/TrustedNodesManager.cs

```diff
@@ -0,0 +1,175 @@
+            }
+            catch (JsonException)
+            {
+                nodes = data.Split(separator, StringSplitOptions.RemoveEmptyEntries);
+            }
+
+            return nodes.Distinct().ToArray();
+        }
+
+        public async Task<bool> AddAsync(string enode, bool updateFile = true)
+        {
+            NetworkNode networkNode = new(enode);
+            if (!_nodes.TryAdd(networkNode.NodeId, networkNode))
+            {
+                if (_logger.IsInfo) _logger.Info($"Trusted node was already added: {enode}");
+                return false;
+            }
+
+            if (_logger.IsInfo) _logger.Info($"Trusted node added: {enode}");
+            Node node = new(networkNode) { IsTrusted = true };
+            NodeAdded?.Invoke(this, new NodeEventArgs(node));
+            if (updateFile)
+            {
+                await SaveFileAsync();
+            }
+
+            return true;
+        }
+
+        public async Task<bool> RemoveAsync(string enode, bool updateFile = true)
+        {
+            NetworkNode networkNode = new(enode);
+            if (!_nodes.TryRemove(networkNode.NodeId, out _))
+            {
+                if (_logger.IsInfo) _logger.Info($"Trusted node was not found: {enode}");
+                return false;
+            }
+
+            if (_logger.IsInfo) _logger.Info($"Trusted node was removed: {enode}");
+            Node node = new(networkNode) { IsTrusted = true };
+            NodeRemoved?.Invoke(this, new NodeEventArgs(node));
+            if (updateFile)
+            {
+                await SaveFileAsync();
+            }
+
+            return true;
+        }
+
+        public bool IsTrusted(string enode)
+        {
+            NetworkNode node = new(enode);
+            return _nodes.ContainsKey(node.NodeId);
+        }
+
+        private Task SaveFileAsync()
+            => File.WriteAllTextAsync(_trustedNodesPath,
+                JsonSerializer.Serialize(_nodes.Values.Select(n => n.ToString()), EthereumJsonSerializer.JsonOptionsIndented));
+
+        public async IAsyncEnumerable<Node> DiscoverNodes([EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            Channel<Node> ch = Channel.CreateBounded<Node>(128);
+
+            foreach (Node node in _nodes.Values.Select(n => new Node(n) { IsTrusted = true }))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                yield return node;
+            }
+
+            void handler(object? _, NodeEventArgs args)
+            {
+                ch.Writer.TryWrite(args.Node);
+            }
+
+            try
+            {
+                NodeAdded += handler;
+
+                await foreach (Node node in ch.Reader.ReadAllAsync(cancellationToken))
+                {
+                    yield return node;
+                }
+            }
+            finally
+            {
+                NodeAdded -= handler;
+            }
```

Comment:
Why do we need a channel?

Reply:
Channel is needed for thread communication, why do you need separate thread here? (IMO you don't)